ir/src/transition_constraints/mod.rs

```diff
@@ -15,57 +15,56 @@ pub const MIN_CYCLE_LENGTH: usize = 2;
 // TRANSITION CONSTRAINTS
 // ================================================================================================
 
-#[derive(PartialEq)]
-enum ConstraintType {
-    Main,
-    Auxiliary,
-}
-
 #[derive(Default, Debug)]
 pub(super) struct TransitionConstraints {
-    /// The indices of the entry nodes for each of the transition constraints against the main trace
-    /// in the graph.
-    main_constraints: Vec<NodeIndex>,
-
-    /// The indices of the entry nodes for each of the transition constraints against the auxiliary
-    /// trace in the graph.
-    aux_constraints: Vec<NodeIndex>,
+    /// Transition constraints against the execution trace, where each index contains a vector of
+    /// the constraint roots for all constraints against that segment of the trace. For example,
+    /// constraints against the main execution trace, which is trace segment 0, will be specified by
+    /// a vector in constraint_roots[0] containing a [NodeIndex] in the graph for each constraint
+    /// against the main trace.
+    constraint_roots: Vec<Vec<NodeIndex>>,
 
     /// A directed acyclic graph which represents all of the transition constraints.
     graph: AlgebraicGraph,
 }
 
 impl TransitionConstraints {
-    // --- PUBLIC ACCESSORS -----------------------------------------------------------------------
+    // --- CONSTRUCTOR ----------------------------------------------------------------------------
 
-    /// Returns a vector of the degrees of the transition constraints.
-    pub fn main_degrees(&self) -> Vec<TransitionConstraintDegree> {
-        self.main_constraints
-            .iter()
-            .map(|entry_index| self.graph.degree(entry_index))
-            .collect()
+    pub fn new(num_trace_segments: TraceSegment) -> Self {
```

Comment:
Same as the previous comment about type of `num_trace_segments` being `TraceSegment`. Probably `usize` is better.